src/test/ui/pattern/or-pattern-macro-pat.rs

```diff
@@ -0,0 +1,40 @@
+// run-pass
+
+//#![feature(or_patterns)]
+
+use Foo::*;
+
+#[derive(Eq, PartialEq, Debug)]
+enum Foo {
+    A(u64),
+    B(u64),
+    C, D
+}
+
+macro_rules! foo {
+    ($orpat:pat, $val:expr) => {
+        match $val {
+            x @ ($orpat) => x,
+            _ => B(0xDEADBEEF),
+        }
+    };
+}
+
+macro_rules! bar {
+    ($orpat:pat, $val:expr) => {
+        match $val {
+            $orpat => 42,
+            _ => 0xDEADBEEF,
+        }
+    };
+}
+
+fn main() {
+    // Test or-pattern.
+    let y = foo!(A(_)|B(_), A(32));
+    assert_eq!(y, A(32));
+
+    // Leading vert in or-pattern.
+    let y = bar!(| C | D, C);
```

Comment:
@petrochenkov For some reason this line is causing a parser failure because the leading `|` is "not expected". However, if I understand the code correctly, the `self.eat_or_separator(None);` should eat the leading vert. Am I misunderstanding?

Reply:
Not sure, from looking at code it seems like it shouldn't fail.

Reply:
Hmm... I wonder if it's because the first and follow sets of :pat change... need to look into this a bit more.

Reply:
The answer is no as far as I can tell. I tried removing `|` from the follow set of `:pat` and it still fails with the same error. I guess that makes sense as, this doesn't seem to be a problem with a metavariable following another...

Reply:
I was right after all. The latest commit adds `|` to the "first set" of the `:pat` non-terminal, which is a breaking change. We'll have to see what crater says...